### PR TITLE
Fix: sourceType not correctly set in AST when provided in options.

### PR DESCRIPTION
**Pull request by armano2**
Even if sourceType is provided to `typescript-estree` its not taken into account and its always doing "autodetection".

Typescript is not using this option anyway, there is no concept of script/module there

fixes: https://github.com/bradzacher/eslint-plugin-typescript/issues/255

### CODE REVIEW

**Comment by j-f1** [comment]
Should this also try to guess a sourceType if one isn’t passed? It should be pretty easy to check if the source contains import or export statements. Also, if they’re present and `sourceType` is `script`, should it throw [like espree does](https://eslint.org/demo/#eyJ0ZXh0IjoiZXhwb3J0IHt9Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJmb3ItZGlyZWN0aW9uIjoyLCJnZXR0ZXItcmV0dXJuIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc29sZSI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Mn0sImVudiI6e319fQ==)?

**Comment by armano2** [comment]
guessing is build in in typescript-estree

File: parser.js [hunk 1]
```diff
@@ -34,6 +34,10 @@ exports.parseForESLint = function parseForESLint(code, options) {
     }
 
     const ast = parse(code, options);
+    if (options.sourceType && ast && ast.type === "Program") {
```
[the remark this answers is not in the document]
Comment (armano2) [reply parser.js]: should we allow autodetection at all? eslint by default sets it to script

[the remark this answers is not in the document]
Comment (j-f1) [reply parser.js]: How does ESLint core do the detection? Do you have to manually specify `sourceType: module`?

[the remark this answers is not in the document]
Comment (armano2) [reply parser.js]: https://eslint.org/docs/user-guide/configuring#specifying-parser-options

- sourceType - set to "script" (default) or "module" if your code is in ECMAScript modules.

[the remark this answers is not in the document]
Comment (armano2) [reply parser.js]: analyzeScope has check like this: 
```js
sourceType: parserOptions.sourceType || "script",
```

[the remark this answers is not in the document]
Comment (armano2) [reply parser.js]: what do you think about handling it in same way?

```js
    if (ast && ast.type === "Program") {
        ast.sourceType = options.sourceType || "script";
    }
```

[the remark this answers is not in the document]
Comment (armano2) [reply parser.js]: but i wanted that this change will not be breaking change, that's why i did it in this way

**Comment by platinumazure** [comment]
Could I get some background on the difference between `options.sourceType` vs `ast.sourceType`?

* Why does the AST (or root node, at any rate) need to have a `sourceType`?
* How is each of these properties used?

**Comment by armano2** [comment]
typescript-estree uses autodetection to determine what is sourceType of parsed file, this data does not come from typescript.

eslint provides `sourceType` field in configuration, in similar way to `jsx`

`options.sourceType` by default in eslint is set to "script" unless overridden to "module". 

i intended to make this change non breaking thats why i added check for that

a lot of plugins are some core rules are using this field to determine behavior of rule:
for example core rule strict.

if we want to release this change with next major update we can cut off fixes for autodetection in `parse.js`

**Comment by JamesHenry** [comment]
@armano2 When you say it doesn't come from TypeScript, I'm not sure exactly what you mean. We convert it from a property in the TypeScript AST like we do with many other things:

```sourceType: node.externalModuleIndicator ? 'module' : 'script'```

**Comment by armano2** [comment]
`externalModuleIndicator` is internal property with array of "importable stuff" and this is not sourceType.

sourceType should be readed from ts config, but we not always have access to it,

in there there is field called module, and there can be alot of values for it:
```
"module": "commonjs",
```

-------

that's why i'm saying this is broken for now, let say i'm compiling everything to commonjs,
and eslint is complaining that i should use module...

**Comment by armano2** [comment]
ok, i updated this PR with changes from typescript-estree

**Pull request merged**